2.2.10 release announcement (#556)
rubygems/bundler-site: source/blog/2021-02-15-a-more-secure-bundler-we-fixed-our-source-priorities.html.markdown

---

title: A more secure bundler: We fixed our source priorities.
date: 2021-02-15 11:33 UTC
tags:
author: David Rodríguez
author_url: https://github.com/deivid-rodriguez/
category: security

---

Last week an article about “Dependency Confusion” hit the news, where a
developer was able to make thousands of dollars on bug bounty programs from big
tech companies, by pushing libraries to public repositories that ended up

> unintentionally being installed into these companies servers.

Reviewer: ```suggestion
unintentionally being installed into these companies' servers.
```

> The developer was able to expose (in a non malicious way) a vulnerability

Reviewer: ```suggestion
The developer was able to expose (in a non-malicious way) a vulnerability
```

> present in well known dependency managers, where given a library name they will

Reviewer: ```suggestion
present in well-known dependency managers, where given a library name they will
```

end up preferring installing it from a public source rather than from a private

> source. This is not secure because the name in the public source is in control

Reviewer: ```suggestion
source. This is not secure because the name in the public source is controlled
```

> of the first person claiming it, whereas the name in the private source is in

Reviewer: ```suggestion
by the first person claiming it, whereas the name in the private source is controlled
```

> control of the private source owner.

Reviewer: ```suggestion
by the private source owner.
```

  private servers is because our system detected them, flagged them for us, and
  we determined them to be non-malicious, and only for research purposes. If
  those gems had been malicious, we wouldn’t have allowed them. Our
  diffend.io friends have published [a more detailed blog post]() about what
  happened in the rubygems.org side of things, and how things were secure.

* We have shipped bundler 2.2.10 with a fix, and now whenever you specify a
  block source in your Gemfile, bundler will prioritize it when resolving direct
  dependencies specified inside, and also transitive dependencies of those. So
  in the following situation both my-private-gem and my-another-private-gem will
  be picked up from https://my-private-server, even if someone pushes a higher
  version with the same name to rubygems.org:

  ~~~ruby
  # my-private-gem.gemspec
  # ...
  gem.dependency("my-another-private-gem")
  ~~~

  ~~~ruby
  # Gemfile

  source "https://rubygems.org"

  source "https://my-private-server" do
    gem "my-private-gem"
  end
  ~~~

As a final note, we’ve been aware of this issue for a while, but unfortunately

> lack resources to take care of everything we need to take care, so the fix was

Reviewer: ```suggestion
lack resources to take care of everything we need to take care of, so the fix was
```

postponed for too long. Maintaining the rubygems.org infrastructure and its
client libraries requires a big amount of work and we barely manage to keep up

> with it. So, if your companies really needs us to stay on top of these issues,

Reviewer: ```suggestion
with it. So, if your company really needs us to stay on top of these issues,
```

please consider funding [RubyTogether](https://rubytogether.org/) :heart:.

That’s all for today,

Happy bundling!

---

> Deivid, Andrè and the Rubygems team

Reviewer: ```suggestion
Deivid, André and the RubyGems team
```

Is that the right way to spell A's name?

Author: I think so!
